sync/src/main/java/tech/pegasys/artemis/sync/SyncManager.java

```diff
@@ -0,0 +1,107 @@
+import tech.pegasys.artemis.networking.eth2.Eth2Network;
+import tech.pegasys.artemis.networking.eth2.peers.Eth2Peer;
+import tech.pegasys.artemis.networking.eth2.rpc.core.InvalidResponseException;
+import tech.pegasys.artemis.networking.eth2.rpc.core.ResponseStream;
+import tech.pegasys.artemis.statetransition.StateTransitionException;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+
+public class SyncManager {
+
+  private final Eth2Network network;
+  private final ChainStorageClient storageClient;
+  private final BlockImporter blockImporter;
+
+  public SyncManager(
+      final Eth2Network network,
+      final ChainStorageClient storageClient,
+      final BlockImporter blockImporter) {
+    this.network = network;
+    this.storageClient = storageClient;
+    this.blockImporter = blockImporter;
+  }
+
+  public CompletableFuture<Void> sync() {
+    Optional<Eth2Peer> possibleSyncPeer = findBestSyncPeer();
+
+    // If there are no peers ahead of us, return
+    if (possibleSyncPeer.isEmpty()) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    Eth2Peer syncPeer = possibleSyncPeer.get();
+    UnsignedLong syncAdvertisedFinalizedEpoch = syncPeer.getStatus().getFinalizedEpoch();
+
+    CompletableFuture<Void> syncCompletableFuture =
+        requestSyncBlocks(syncPeer, blockResponseListener(syncPeer));
+
+    return syncCompletableFuture.thenRun(
+        () -> {
+          if (storageClient.getFinalizedEpoch().compareTo(syncAdvertisedFinalizedEpoch) < 0) {
+            disconnectFromPeerAndRunSyncAgain(syncPeer);
+          }
+        });
+  }
+
+  private Optional<Eth2Peer> findBestSyncPeer() {
+    UnsignedLong ourFinalizedEpoch = storageClient.getFinalizedEpoch();
+    return network
+        .streamPeers()
+        .filter(peer -> peer.getStatus().getFinalizedEpoch().compareTo(ourFinalizedEpoch) > 0)
+        .max(Comparator.comparing(p -> p.getStatus().getFinalizedEpoch()));
+  }
+
+  private CompletableFuture<Void> requestSyncBlocks(
+      Eth2Peer peer, ResponseStream.ResponseListener<BeaconBlock> blockResponseListener) {
+    Eth2Peer.StatusData peerStatusData = peer.getStatus();
+    Bytes32 headBlockRoot = peerStatusData.getHeadRoot();
+    UnsignedLong headBlockSlot = peerStatusData.getHeadSlot();
+    UnsignedLong startSlot = compute_start_slot_at_epoch(storageClient.getFinalizedEpoch());
+    UnsignedLong step = UnsignedLong.ONE;
+    UnsignedLong count = headBlockSlot.minus(startSlot);
+    return peer.requestBlocksByRange(headBlockRoot, startSlot, count, step, blockResponseListener);
+  }
+
+  private ResponseStream.ResponseListener<BeaconBlock> blockResponseListener(Eth2Peer peer) {
+    return ((block) -> {
+      try {
+        blockImporter.importBlock(block);
+      } catch (StateTransitionException e) {
+        disconnectFromPeerAndRunSyncAgain(peer);
+        throw new InvalidResponseException("Received bad block from peer", e);
```

Comment:
If we throw here, the failure should propagate back up to `syncCompletableFuture`.  We could chain `disconnectFromPeerAndRunSyncAgain` onto that future ...